tidy3d/components/boundary.py

```diff
@@ -0,0 +1,626 @@
+    )
+
+    parameters: PMLParams = pd.Field(
+        DefaultStablePMLParameters,
+        title="Stable PML Parameters",
+        description="'Stable' parameters of the complex frequency-shifted absorption poles.",
+    )
+
+
+class Absorber(AbsorberSpec):
+    """Specifies an adiabatic absorber along a single dimension.
+    This absorber is well-suited for dispersive materials
+    intersecting with absorbing edges of the simulation at the expense of more layers.
+
+    Example
+    -------
+    >>> pml = Absorber(num_layers=40)
+    """
+
+    num_layers: pd.NonNegativeInt = pd.Field(
+        40,
+        title="Number of Layers",
+        description="Number of layers of absorber to add to + and - boundaries.",
+    )
+
+    parameters: AbsorberParams = pd.Field(
+        DefaultAbsorberParameters,
+        title="Absorber Parameters",
+        description="Adiabatic absorber parameters.",
+    )
+
+
+# pml types allowed in simulation init
+PMLTypes = Union[PML, StablePML, Absorber, None]
+
+
+# """ boundary specification classes """
+
+# types of boundaries that can be used in Simulation
+
+Symmetry = Literal[0, -1, 1]
+BoundaryEdgeType = Union[
+    Periodic, PECBoundary, PMCBoundary, PML, StablePML, Absorber, BlochBoundary
+]
+
+
+class Boundary(Tidy3dBaseModel):
+    """Boundary conditions at the minus and plus extents along a dimension
+
+    Example
+    -------
+    >>> boundary = Boundary(plus = PML(), minus = PECBoundary())
+    """
+
+    plus: BoundaryEdgeType = pd.Field(
+        Periodic(),
+        title="Plus BC",
+        description="Boundary condition on the plus side along a dimension.",
+    )
+
+    minus: BoundaryEdgeType = pd.Field(
+        Periodic(),
+        title="Minus BC",
+        description="Boundary condition on the minus side along a dimension.",
+    )
+
+    @pd.root_validator
+    def bloch_on_both_sides(cls, values):
+        """Error if a Bloch boundary is applied on only one side."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+        num_bloch = isinstance(plus, BlochBoundary) + isinstance(minus, BlochBoundary)
+        if num_bloch == 1:
+            raise SetupError(
+                "Bloch boundaries must be applied either on both sides or on neither side."
+            )
+        return values
+
+    @pd.root_validator
+    def periodic_with_pml(cls, values):
+        """Error if PBC is specified with a PML."""
+        plus = values.get("plus")
+        minus = values.get("minus")
+        num_pbc = isinstance(plus, Periodic) + isinstance(minus, Periodic)
+        num_pml = isinstance(plus, (PML, StablePML, Absorber)) + isinstance(
+            minus, (PML, StablePML, Absorber)
+        )
+        if num_pbc == 1 and num_pml == 1:
+            raise SetupError("Cannot have both PML and PBC along the same dimension.")
+        return values
+
+    @pd.root_validator
+    def periodic_with_pec_pmc(cls, values):
```

Comment:
I wonder if this should happen internally (on server-side processing), might be confusing if someone sets `minus=Periodic` but then realizes later it's set to `PECBoundary`.  not sure how important it is either way, up to you.

Reply:
How about printing a warning when this happens? This validation will later affect the size of the grid (due to halos) and the way fields are computed at monitors close to a domain edge, so having a discrepancy between the frontend and backend boundary definitions might come back to haunt us.

Reply:
ok, we can do warning then.